### DISCUSSION TITLE
commitlint.sh (skip) not specified in config file

### DISCUSSION THREAD

**Discussion by sbsrnt** [evilmartians/lefthook#639] [Q&A] [answered]
I created a package that has shared `lefthook` config: https://github.com/exile-watch/splinters/tree/main/packages/lefthook-config
The file tree of above package:
```
root // @exile-watch/splinters lerna monorepo
| - .lefthook/commit-msg/commitlint.sh
| - lefthook.yml
| - packages
    | - lefthook-config
        | - lefthook.yml # shareable config
        | - scripts
            | - commitlint.sh
```

The root `lefthook.yml` looks like [this](https://github.com/exile-watch/splinters/blob/main/lefthook.yml):
```yaml
remotes:
  - git_url: https://github.com/exile-watch/splinters
    configs:
      - packages/lefthook-config/lefthook.yml
```
Now, for some reason `commit-msg` script is working here.

But when I try to apply the same settings (literally copy-paste) to another repo it doesn't seem to find `commitlint.sh`: https://github.com/exile-watch/nucleus

On top of that, I'm not sure why the output lists `commitlint.sh (skip)` twice:
```
╭──────────────────────────────────────╮
│ 🥊 lefthook v1.6.1  hook: commit-msg │
╰──────────────────────────────────────╯
│  commitlint.sh (skip) not specified in config file
│  commitlint.sh (skip) not specified in config file
```
---
I thought I understood how to create a shareable solution but now I'm confused why `commitlint.sh` is not specified ._.
And why the script is specified in `@exile-watch/splinters` and not in `@exile-watch/nucleus` even though both root `lefthook.yml` rely on `remotes` config

**Comment by mrexox** [comment, marked as answer]
Hey! I the problem is that the script file name is [`commitlint.sh`](https://github.com/exile-watch/splinters/blob/main/.lefthook/commit-msg/commitlint.sh) but the config has [`commitlint`](https://github.com/exile-watch/splinters/blob/main/packages/lefthook-config/lefthook.yml#L18) instead. To fix this the config should be changed to the following:

```yml
commit-msg:
  scripts:
    "commitlint.sh":
      runner: bash
```

Or the script name should be changed to just `commitlint`.

**Reply by sbsrnt** [reply]
Yeah that seemed to be the case, quite silly of me 😅
For brevity I will stick with `.sh` extensions for scripts as that seemed to be slightly confusing with `commands`.

---

For the second part, do you reckon what could be the cause of scripts triggering twice 🤔?

**Reply by mrexox** [reply]
Not sure about the script triggering twice 🤔 I couldn't reproduce it locally. If you face this issue again, please come back with the reproduction instructions 🙏

**Reply by sbsrnt** [reply]
yeah I wonder if bash has any role in it, all cases are related to `.lefthook/commit-msg/commitlint.sh`

---
case 1 (with `run_commitlint`):
```
#!/bin/bash
source ./node_modules/@exile-watch/lefthook-config/scripts/commitlint.sh

run_commitlint
```

outputs:

```
╭──────────────────────────────────────╮
│ 🥊 lefthook v1.6.1  hook: commit-msg │
╰──────────────────────────────────────╯
┃  commitlint.sh ❯

⧗   input: bricked commit
✖   subject may not be empty [subject-empty]
✖   type may not be empty [type-empty]

✖   found 2 problems, 0 warnings
ⓘ   Get help: https://docs.exile.watch/development/commit-message-guidelines


exit status 1┃  commitlint.sh ❯

⧗   input: bricked commit
✖   subject may not be empty [subject-empty]
✖   type may not be empty [type-empty]

✖   found 2 problems, 0 warnings
ⓘ   Get help: https://docs.exile.watch/development/commit-message-guidelines

```

---

case 2 (without `run_commitlint`):

```
#!/bin/bash
source ./node_modules/@exile-watch/lefthook-config/scripts/commitlint.sh
```

outputs

```
╭──────────────────────────────────────╮
│ 🥊 lefthook v1.6.1  hook: commit-msg │
╰──────────────────────────────────────╯
┃  commitlint.sh ❯ 


┃  commitlint.sh ❯ 

⧗   input: bricked commit
✖   subject may not be empty [subject-empty]
✖   type may not be empty [type-empty]

✖   found 2 problems, 0 warnings
ⓘ   Get help: https://docs.exile.watch/development/commit-message-guidelines


exit status 1
  ────────────────────────────────────
summary: (done in 1.60 seconds)
✔️  commitlint.sh
🥊  commitlint.sh
```

---

case 3 (only `run_commitlint`)
```
#!/bin/bash
run_commitlint
```
outputs
````
C:\Users\seb\WebstormProjects\nucleus\.lefthook\commit-msg\commitlint.sh: line 2: run_commitlint: command not found
┃  commitlint.sh ❯


exit status 127┃  commitlint.sh ❯ 

⧗   input: bricked commit
✖   subject may not be empty [subject-empty]
✖   type may not be empty [type-empty]

✖   found 2 problems, 0 warnings
ⓘ   Get help: https://docs.exile.watch/development/commit-message-guidelines


exit status 1
  ────────────────────────────────────
summary: (done in 1.55 seconds)
🥊  commitlint.sh
🥊  commitlint.sh

**Reply by sbsrnt** [reply]
Removing `.lefthook/commit-msg/commitlint.sh` removed the duplicated output 🤔

**Reply by sbsrnt** [reply]
I went with:
1. Remove `.lefthook/commit-msg/commitlint.sh`
2. `git commit -a -m "bricked commit" -> received only 1 errored output (correct)
3. `npx lefthook uninstall` 
4. `npx lefthook install`
5. `git commit -a -m "bricked commit" -> received only 1 errored output (correct)

Now I'm double confused if `.lefthook` dir is required lol

**Reply by sbsrnt** [reply]
Now I also:
1. removed `.git/hooks` dir
2. `npx lefthook uninstall`
Error: open .git/hooks: The system cannot find the file specified. (makes sense)
3. `npx lefthook install`
4. `git commit -a -m "bricked commit" -> received only 1 errored output (correct)

All the steps above were done without creating `.lefthook/*` dir - I guess that's... alright? @mrexox 

Even though it's slightly contradictory to the [script docs](https://github.com/evilmartians/lefthook/blob/master/docs/configuration.md#source_dir) 🤔not sure if `remotes` and [scripts defined in repo](https://github.com/exile-watch/splinters/tree/main/packages/lefthook-config) we are importing from are playing any role into that but it works - question is if it's a good thing or not

**Reply by mrexox** [reply]
So, you see the error twice because you have `.lefthook/commit-msg/commitlint.sh` both locally and in one of the remotes. I am not sure about how to handle this case 🤔 It's probably a very unusual case, but I will put it on my radar to think this out. Thank you for explaining the steps to reproduce it.

**Reply by sbsrnt** [reply]
I mean to be fair... wouldn't consuming remote `.lefthook` dir (like it's doing now) be a cleaner approach then 🤔? 
(assumig we utilize `remotes` part in local `lefthook.yml` 

Creating only `lefthook.yml` in a repo that consumes shared `lefthook-config` package with this content
```
remotes:
  - git_url: <git url>
    configs:
      - packages/lefthook-config/lefthook.yml
```
sounds pretty straight forward and is one less directory to maintain

---

And I suppose `.lefthook/*` dir in consuming repo would overwrite remote `.lefthook/*` scripts (or whatever is in it) (if it's not doing that already)
